tla-types/src/main/scala/at/forsyte/apalache/tla/typecheck/etc/ToEtcExpr.scala

```diff
@@ -197,6 +197,16 @@ class ToEtcExpr(annotationStore: AnnotationStore, aliasSubstitution: ConstSubsti
       mkApp(ref, Seq(sig), args.map(this(_)): _*)
     }
 
+    def diffArgTypes(args: List[TlaEx], expectedTypes: List[TlaType1], actualTypes: List[TlaType1]): List[String] = {
+      val illTypedArguments = args.zip(expectedTypes).zip(actualTypes).filter { case ((_, expectedType), argType) =>
+        expectedType != argType
+      }
+
+      illTypedArguments.map { case ((arg, expectedType), argType) =>
+        s"Argument $arg should have type $expectedType but has type $argType"
+      }
+    }
```

Comment:
Also, not too important here, but this may be helpful later: if you're going to do a map after a filter, Scala has [`withFilter`](https://stackoverflow.com/questions/19617378/withfilter-instead-of-filter) which is more efficient when the collection is large.